sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+            new RestClientBuilder.HttpClientConfigCallback() {
+              public HttpAsyncClientBuilder customizeHttpClient(
+                  HttpAsyncClientBuilder httpAsyncClientBuilder) {
+                return httpAsyncClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+              }
+            });
+      }
+      return restClientBuilder.build();
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} reading data from Elasticsearch.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<String>> {
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    @Nullable abstract String getQuery();
+    @Nullable abstract String getScrollKeepalive();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+      abstract Builder setQuery(String query);
+      abstract Builder setScrollKeepalive(String scrollKeepalive);
+      abstract Read build();
+    }
+
+    public Read withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.read()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Read withQuery(String query) {
+      checkArgument(query != null, "ElasticsearchIO.read().withQuery(query) called with null "
+          + "query");
+      return builder().setQuery(query).build();
+    }
+
+    public Read withScrollKeepalive(String scrollKeepalive) {
+      checkArgument(scrollKeepalive != null, "ElasticsearchIO.read().withScrollKeepalive"
+          + "(keepalive) called with null keepalive");
+      return builder().setScrollKeepalive(scrollKeepalive).build();
+    }
+
+    @Override
+    public PCollection<String> apply(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new BoundedElasticsearchSource(this, null)
+      ));
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.read() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      getConnectionConfiguration().populateDisplayData(builder);
+    }
+
+  }
+
+  /**
+   * A {@link BoundedSource} reading from Elasticsearch.
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
+
+    private ElasticsearchIO.Read spec;
+    @Nullable
+    private final String shardPreference;
```

Comment:
Document this? (the word "preference" makes me worry - does ES guarantee that reading with a different "shardPreference" is guaranteed to give non-overlapping results?)